### PR TITLE
Feature/spk 142. Limit image size while NFC key have enough space to store it

### PR DESCRIPTION
**Pull request by AlexMisiulia**

### CODE REVIEW

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt [hunk 1]
```diff
@@ -87,11 +94,43 @@ class WriteNfcViewModel @Inject constructor(
...
+
+    private suspend fun getPhotoByteArray(
+        nfcNdefData: NfcNdefData
+    ): ByteArray? {
+        val nfcKeyCapacity = nfcInteractor.getMaxSize(nfcNdefData)
+        val nfcPersonalFormDataSize = nfcInteractor.getConvertedSize(nfcPersonalFormData)
+
+        val maxPhotoSize = if (nfcKeyCapacity != null && nfcPersonalFormDataSize != null) {
+            nfcKeyCapacity - nfcPersonalFormDataSize
+
```
Comment (AlexMisiulia) [inline app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: Remove blank line

File: common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt [hunk 2]
```diff
@@ -96,12 +98,35 @@ class ImageUtils @Inject constructor(private var ioDispatcher: CoroutineDispatch
...
+
+            do {
+                outputStream = ByteArrayOutputStream()
+                logD(TAG, "limit photo size, currQuality = $currQuality")
+                bitmap.compress(Bitmap.CompressFormat.JPEG, currQuality, outputStream)
+                currSize = outputStream.toByteArray().size
+                currQuality -= qualityDecreaseStep
+
+                // can't decode the image to the maxDiskSpace
+                if(currQuality <= 0) return@withContext null
```
Comment (AlexMisiulia) [inline common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt]: Move to the loop beginning

File: app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModelTest.kt [hunk 3]
```diff
@@ -5,29 +5,57 @@ import com.yoti.mobile.android.social.common.features.nfc.data.NfcWriteException
...
+    )
     private lateinit var loadingObserver: LiveDataTestObserver<Boolean>
 
     @Before
     fun before() {
-        loadingObserver = LiveDataTestObserver(viewModel.isLoadingLiveData)
+        runBlocking {
+            loadingObserver = LiveDataTestObserver(viewModel.isLoadingLiveData)
+            whenever(repository.imagePath).thenReturn(PHOTO_PATH)
+
```
[the remark this answers is not in the document]
Comment (AlexMisiulia) [reply app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModelTest.kt]: Done

File: app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModelTest.kt [hunk 4]
```diff
@@ -133,6 +161,66 @@ class WriteNfcViewModelTest : BaseNfcViewModelTest<WriteNfcViewModel>() {
...
+        // act
+        viewModel.onNfcEvent(nfcNdefData)
+
+        // assert
+        val result = viewModel.errorDialog.value?.peekContent()
+        assertEquals(DialogInfo(R.string.write_key_not_enough_space_error_msg, false), result)
+    }
+
+    @Test
+    fun `write photo reducing its size to nfc key capacity`() = runBlocking {
```
[the remark this answers is not in the document]
Comment (AlexMisiulia) [reply app/src/test/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModelTest.kt]: Done

File: common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt (hunk 2 above)
Comment (AlexMisiulia) [reply common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt (hunk 1 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt [hunk 5]
```diff
@@ -87,11 +94,43 @@ class WriteNfcViewModel @Inject constructor(
...
             setNfcError(nfcWriteException)
             false
         }
     }
+
+    private suspend fun getPhotoByteArray(
+        nfcNdefData: NfcNdefData
+    ): ByteArray? {
+        val nfcKeyCapacity = nfcInteractor.getMaxSize(nfcNdefData)
+        val nfcPersonalFormDataSize = nfcInteractor.getConvertedSize(nfcPersonalFormData)
```
[the remark this answers is not in the document]
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: Done

File: common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt (hunk 2 above)
Comment (AlexMisiulia) [reply common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt]: Done

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt [hunk 6]
```diff
@@ -42,13 +46,16 @@ class WriteNfcViewModel @Inject constructor(
...
-            } else {
-                setLoading(false)
+            val imagePath = personalFormRepository.imagePath
+            setLoading(false)
+
+            if (personalForm != null && imagePath != null) {
                 nfcPersonalFormData = NfcPersonalFormData(personalForm, pinCode)
+                photoPath = imagePath
+            } else {
+                finishScreen()
```
Comment (cquemin) [inline app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: Why is this in the VM ? it should be in the view in response to a `LiveData` no?

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt [hunk 7]
```diff
@@ -87,11 +94,42 @@ class WriteNfcViewModel @Inject constructor(
...
+
+            nfcInteractor.writeAsync(nfcNdefData, nfcPersonalFormData.copy(personalForm = personalFormWithPhoto))
             true
         } catch (nfcWriteException: NfcWriteException) {
             setNfcError(nfcWriteException)
             false
         }
     }
+
+    private suspend fun getPhotoByteArray(
```
Comment (cquemin) [inline app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: This suspended function should return a bitmap I would think. So you don't have to deserialise a bitmap on the UI thread.

File: common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt [hunk 8]
```diff
@@ -96,12 +98,36 @@ class ImageUtils @Inject constructor(private var ioDispatcher: CoroutineDispatch
         BitmapFactory.decodeByteArray(photoBytes, 0, photoBytes.size)
     }
 
-    suspend fun toByteArray(imagePath: String, maxWidth: Int, maxHeight: Int, quality: Int = 100): ByteArray =
-        withContext(ioDispatcher) {
+    suspend fun toByteArray(
```
Comment (cquemin) [inline common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt]: Same question here, should this return a `Bitmap`?

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt (hunk 6 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: This method sets live data under the hood:

```kotlin
    protected fun finishScreen() {
        _finishScreenLiveData.value = Event(true)
    }
```

Maybe the name confuses you. Do you prefer `setFinishScreen(true)`?

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt (hunk 7 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: This performs before writing the data to NFC key. I wouldn't convert the photo to Bitmap after that, I will write `ByteArray` straight to the NFC key.

File: common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt (hunk 8 above)
Comment (AlexMisiulia) [reply common/src/main/java/com/yoti/mobile/android/social/common/utils/ImageUtils.kt]: As per comment above

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt (hunk 6 above)
Comment (cquemin) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: Yeah probably that would be better and more consistent with what you have done so far.

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt (hunk 7 above)
Comment (cquemin) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: Fair enough

File: app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt (hunk 6 above)
Comment (AlexMisiulia) [reply app/src/main/java/com/yoti/mobile/android/social/keys/features/writekey/WriteNfcViewModel.kt]: Done

**Pull request merged**